packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -29,64 +38,165 @@ class UpgradeCommand extends FlutterCommand {
+      if (flutterVersion.channel != 'master' && FlutterVersion.officialChannels.contains(flutterVersion.channel)) {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. It is recommended to use git directly if not working off of '
+          'an official channel.'
+        );
+      // Otherwise explain that local changes can be lost.
+      } else {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. If it is okay to remove local changes, then re-run this '
+          'command with --force.'
+        );
+      }
+    }
+    final String stashName = await maybeStash(gitTagVersion);
+    await upgradeChannel(flutterVersion);
+    await attemptRebase();
+    await precacheArtifacts();
+    await updatePackages(flutterVersion);
+    await runDoctor();
+    await applyStash(stashName);
+    return null;
+  }
+
+  /// Check if there is an upstream repository configured.
+  ///
+  /// Exits tool if there is no upstream.
+  Future<void> verifyUpstreamConfigured() async {
     try {
       await runCheckedAsync(<String>[
         'git', 'rev-parse', '@{u}',
       ], workingDirectory: Cache.flutterRoot);
     } catch (e) {
       throwToolExit('Unable to upgrade Flutter: no upstream repository configured.');
     }
+  }
 
-    final FlutterVersion flutterVersion = FlutterVersion.instance;
+  /// Attempt to stash any local changes.
+  ///
+  /// Returns the stash name if any changes were stashed. Exits tool if
+  /// `git stash` returns a non-zero exit code.
+  Future<String> maybeStash(GitTagVersion gitTagVersion) async {
+    final String stashName = 'flutter-upgrade-from-v${gitTagVersion.x}.${gitTagVersion.y}.${gitTagVersion.z}';
+    try {
+      final RunResult runResult = await runCheckedAsync(<String>[
+        'git', 'stash', 'push', '-m', stashName
+      ]);
+      // output message will contain stash name if any changes were stashed..
+      if (runResult.stdout.contains(stashName)) {
+        return stashName;
+      }
+    } catch (e) {
+      throwToolExit('Failed to stash local changes: $e');
+    }
+    return null;
+  }
 
+  /// Attempts to upgrade the channel.
+  ///
+  /// If the user is on a deprecated channel, attempts to migrate them off of
+  /// it.
+  Future<void> upgradeChannel(FlutterVersion flutterVersion) async {
     printStatus('Upgrading Flutter from ${Cache.flutterRoot}...');
-
     await ChannelCommand.upgradeChannel();
+  }
 
-    int code = await runCommandAndStreamOutput(
-      <String>['git', 'pull', '--ff-only'],
+  /// Attempts to rebase the upstream onto the local branch.
+  ///
+  /// If there haven't been any hot fixes or local changes, this is equivalent
+  /// to a fast-forward.
+  Future<void> attemptRebase() async {
+    final int code = await runCommandAndStreamOutput(
+      <String>['git', 'pull', '--rebase'],
```

Comment:
Should we be checking if the rebase completed successfully, and if not, issue a `git rebase --abort`? We're mitigating some of that with the stash which is good, but I suspect a hot fix could potentially put someone in this state if anything with the hotfix process goes wrong - leaving them with an unusable installation.

I'm not sure if git sets an exit code in this case. Maybe the answer is to advise them to completely reinstall flutter instead.

Reply:
will a rebase work if the commits we put on the hot fix branch weren't on the new branch?

Reply:
If we checkout the hotfix branch from the same commit that the dev/beta/stable was picked from, then that would have the same history +1 additional commit and the rebase or ff would work.

If they upgrade and try to leave the hot fix branch to get back to something based off of master, now there are conflicts because the hotfix commit. We would actually have to reset/checkout to the commit we branched the hotfix from and then fast forward from there?

Reply:
would that be a reset?

cc @Hixie @tvolkert

Reply:
I think if we do a `git stash (labeled or whatever) && git fetch upstream && git reset --hard upstream/branchname && git stash pop` we should be good.

Rebase has the potential to mess you up.

Reply:
The git stash pop can mess you up too, but only if you've modified the flutter repo, in which case you should hopefully know how to modify it again (but this could be something like Xcode changed some properties on flutter_gallery or something).

Reply:
I think we could reset to the tag name the hotfix was branched from right? that would have the fewest changes before the ff

Reply:
The idea here is that people will be on one branch, e.g. `stable`, the whole time, and the commits on that branch will change over time (and not as strict supersets each time). For example:
* v3.4.5 tag is made, points to a commit on master.
* stable branch is made to point to v3.4.5.
* v3.4.5-hotfixes branch is made, forking master from v3.4.5.
* a commit X is added only to v3.4.5-hotfixes branch.
* v3.4.5-hotfix.1 tag is made, points to X.
* stable branch point is made to point to v3.4.5-hotfix.1.
* v3.7.6 tag is made, points to a commit on master. History of this commit does not include X.
* stable branch point is made to point to v3.7.6. History of this commit does not include X.

A user which was on stable v3.4.5-hotfix.1 will now attempt to upgrade to v3.7.6 in one step, without changing branch. I would expect rebasing, merging, etc, to fail to do this successfully. Only a forced reset seems likely to succeed at doing this.

Reply:
In https://github.com/flutter/flutter/pull/29786, I've updated this to first check if we're on a hotfix branch (v3.4.5-hotfix.1). If so, I do a `git reset --hard v3.4.5`, then `ff` as normal